File: Adaptive_Filter/vhls/step2/nlms_filter.cpp

#include "nlms_filter.h"

nlms_filter::nlms_filter()
{
	// Just call superclass constructor
}

void nlms_filter::update_coefs(short* coefs, short* noise_buf, short step_size, short err)
{
	int   ii            = 0;
	int   noise_buf_idx = NUM_COEFS-1;
	long  tmp           = 0;
	long  tmp2          = 0;
	int   l2norm        = 0;

	// Compute L2 norm of the noise buffer
	for (ii = 0; ii < NUM_COEFS; ii++)
	{
		l2norm += (int)noise_buf[ii] * (int)noise_buf[ii];
	}
	l2norm >>= sizeof(short)*8;

	for (ii = 0; ii < NUM_COEFS; ii++)
	{

		// Update equation
		tmp = (long)step_size * (long)err * (long)noise_buf[noise_buf_idx]; // This node is 2x data_width + coef_width
		tmp /= l2norm;
		tmp = tmp >> (sizeof(short)*8-2);

		// Saturate output if coefficients get too big
		tmp2 = (long)coefs[ii] + tmp;
		if (tmp2 > (1 << ((sizeof(short)*8-1)-1)))
			coefs[ii] = (1 << ((sizeof(short)*8-1)-1));
		else if (tmp2 < -(1 << (sizeof(short)*8-1)))
			coefs[ii] = -(1 << (sizeof(short)*8-1));
		else
			coefs[ii] = tmp2;

		// Increment noise buffer
		noise_buf_idx = noise_buf_idx - 1;
	}
}

nlms_filter::~nlms_filter()
{
	// Just call superclass destructor
}

File: Adaptive_Filter/matlab/model/nlms_filter.m

classdef nlms_filter < adaptive_filter

	methods (Access = public)
		function obj = nlms_filter()
			obj = obj@adaptive_filter();
		end	
	end
	
	methods (Access = protected)
		function coefs = update_coefs(obj, coefs, noise_buf, step_size, leakage, err)
		
            % Compute L2 Norm
            l2norm = 0;
            for ii = 1:length(noise_buf)
                l2norm = l2norm + noise_buf(ii) * noise_buf(ii);
            end
            
			noise_buf_idx = length(noise_buf);
			for ii = 1:length(coefs)
				coefs(ii) = coefs(ii) + step_size * err * noise_buf(noise_buf_idx) / l2norm;
				noise_buf_idx = noise_buf_idx - 1;
			end
			
		end
		
	end

end